packages/demux/src/AloxideActionHandler.ts

```diff
@@ -1,151 +1,158 @@
-      });
-  }
-
-  protected async handleWithState(handle: (state: any, context?: any) => void): Promise<void> {
-    await handle(this.state);
-
-    const { blockNumber } = this.indexStateModel;
-
-    this.stateHistory[blockNumber] = JSON.parse(JSON.stringify(this.state));
-
-    if (
-      blockNumber > this.stateHistoryMaxLength &&
-      this.stateHistory[blockNumber - this.stateHistoryMaxLength]
-    ) {
-      delete this.stateHistory[blockNumber - this.stateHistoryMaxLength];
-    }
-  }
-
-  protected async setup(): Promise<void> {
-    /**
-     * Table DemuxIndexState
-     */
-    if (this.dataAdapter.setup) {
-      return this.dataAdapter.setup(this.getIndexStateModelName());
-    }
-  }
-
-  protected async rollbackTo(blockNumber: number): Promise<void> {
-    const latestBlockNumber = this.indexStateModel.blockNumber;
-
-    const toDelete = [...Array(latestBlockNumber - blockNumber).keys()].map(
-      n => n + blockNumber + 1,
-    );
-
-    for (const n of toDelete) {
-      delete this.stateHistory[n];
-    }
-
-    this.state = this.stateHistory[blockNumber];
-  }
-}
+import { EntityConfig } from '@aloxide/bridge/src';
+import { AbstractActionHandler } from 'demux';
+
+import { indexStateSchema } from './indexStateSchema';
+
+import type {
+  NextBlock,
+  HandlerVersion,
+  IndexState,
+  ActionHandlerOptions,
+  HandlerInfo,
+} from 'demux';
+import type { DataAdapter } from './DataAdapter';
+import type { DMeta } from './DMeta';
+import type { IndexStateModel } from './IndexStateModel';
+
+export interface AloxideActionHandlerOptions extends ActionHandlerOptions {
+  indexStateModelName?: string;
+}
+
+export interface AloxideActionHandlerContext {
+  info: HandlerInfo;
+}
+
+export class AloxideActionHandler extends AbstractActionHandler {
+  indexStateModel: IndexStateModel;
+  indexStateEntity: EntityConfig = indexStateSchema;
+
+  protected indexStateModelName: string;
+
+  constructor(
+    protected bcName: string,
+    protected dataAdapter: DataAdapter<any, any>,
+    handlerVersions: HandlerVersion[],
+    options?: AloxideActionHandlerOptions,
+  ) {
+    super(handlerVersions, options);
+    if (options) {
+      this.indexStateModelName = options.indexStateModelName;
+    }
+  }
+
+  getIndexStateModelName() {
+    return this.indexStateModelName || `DemuxIndexState_${this.bcName.replace(/\W+/, '_')}`;
+  }
+
+  protected updateIndexState(
+    state: any,
+    nextBlock: NextBlock,
+    isReplay: boolean,
+    handlerVersionName: string,
+    context?: AloxideActionHandlerContext,
+  ): Promise<void> {
+    const block = nextBlock.block;
+
+    this.indexStateModel.blockNumber = block.blockInfo.blockNumber;
+    this.indexStateModel.blockHash = block.blockInfo.blockHash;
+    this.indexStateModel.isReplay = isReplay;
+    this.indexStateModel.handlerVersionName = handlerVersionName;
+
+    if (state) {
+      this.indexStateModel.state = JSON.stringify(state);
```

Comment:
Is it right to alway store state in the index-state? We might need an option for disabling this.

Reply:
continue in #77